### PR TITLE
fix(object-file): allow boxed primitives, Set, Map and RegExp in values

### PR DESCRIPTION
**Pull request by tinovyatkin**
With current implementation ObjectFile (and all it subclasses) are throwing on encountering any perfectly valid Javascript boxed primitives in the object values (for example `new Sting('something')` or `new Boolean(something)`) even while those values may be stringified by `JSON.stringify` (the serialisation function that we are using) without any problem. This PR fixes that.

In addition, there are several JavaScript built-in objects that while having some limitations on JSII are very useful either for internal usage or for TS/JS-first usage of projen, however, those classes are missing built-in `.toJSON` method and so also resulting in throw from ObjectFile, particularly RegExp, Set and Map. This PR proposes to fix that by resolving RegExp (without flags) into its string source, Set into an array, and Map into an object. Note, that if we have a subclass of any of those built-in types that implements custom `.toJSON` then it will continue to work as with current version (see added tests).

As a simple demonstration of value of stringifiying `Set` I've included a small commit into this PR that switches implementation of ESLint fields that are expected to have non-repeated values (`plugins` and `extends`) from Array to Set using the fix implemented in this PR.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/_resolve.ts [hunk 1]
```diff
@@ -17,6 +18,23 @@ export function resolve(value: any, options: ResolveOptions = {}): any {
     return resolve(resolved, options);
   }
 
+  // Special resolution for few JavaScript built-in types
+  // that by default would be stringified as empty objects ('{}')
+  // as they are missing a `toJSON` implementation.
+  switch (true) {
+    case types.isRegExp(value) && !value.flags:
+      return value.source;
```
Comment (mrgrain) [inline src/_resolve.ts:26]: I think this is making to many assumptions. Specifically it is a different implementation then [RegExp.prototype.toString()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/toString).

Happy to accept an implementation using `toString` instead. But also it seems okay to me to expect a user to call `toString()` themselves. For the Set & Map the conversion is less straightforward, so I agree the benefits outweigh the ambiguity.

**Review by mrgrain — CHANGES_REQUESTED** [review]

File: src/_resolve.ts (hunk 1 above)
Comment (mrgrain) [reply src/_resolve.ts:26]: Yes, hence my reluctance. In ESLint and Jest these are just strings that happen to be used as a RegExp. 
`/^tada$/.source` is equally easy to write.

Comment (tinovyatkin) [reply src/_resolve.ts:26]: `/^tada$/` seems to be even easier to write. But right now just throws with very very cryptic error message. I can keep that behaviour as is if you believe that is better DX than I'm proposing in this PR.

Comment (mrgrain) [reply src/_resolve.ts:26]: 🤦🏻 Of course. Can we throw with a better error in that case?

**Review by mrgrain — APPROVED** [review]
Sorry this took so long. Approved and should be good to ship once the conflict is resolved.

**Pull request merged**